plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessExtension.java

```diff
@@ -201,14 +224,66 @@ public void setEnforceCheck(boolean enforceCheck) {
 		} else {
 			try {
 				Constructor<T> constructor = clazz.getConstructor(SpotlessExtension.class);
-				T newlyCreated = constructor.newInstance(this);
-				formats.put(name, newlyCreated);
-				return newlyCreated;
+				T formatExtension = constructor.newInstance(this);
+				formats.put(name, formatExtension);
+				createFormatTask(name, formatExtension);
+				return formatExtension;
 			} catch (NoSuchMethodException e) {
 				throw new GradleException("Must have a constructor " + clazz.getSimpleName() + "(SpotlessExtension root)", e);
 			} catch (Exception e) {
 				throw Errors.asRuntime(e);
 			}
 		}
 	}
+
+	@SuppressWarnings("rawtypes")
+	private void createFormatTask(String name, FormatExtension formatExtension) {
+		// create the SpotlessTask
+		String taskName = EXTENSION + SpotlessPlugin.capitalize(name);
+		SpotlessTask spotlessTask = project.getTasks().create(taskName, SpotlessTask.class);
+		project.afterEvaluate(unused -> formatExtension.setupTask(spotlessTask));
+
+		// clean removes the SpotlessCache, so we have to run after clean
+		Task clean = project.getTasks().getByName(BasePlugin.CLEAN_TASK_NAME);
+		spotlessTask.mustRunAfter(clean);
+
+		// create the check and apply control tasks
+		Task checkTask = project.getTasks().create(taskName + CHECK);
+		Task applyTask = project.getTasks().create(taskName + APPLY);
+
+		checkTask.dependsOn(spotlessTask);
+		applyTask.dependsOn(spotlessTask);
+		// when the task graph is ready, we'll configure the spotlessTask appropriately
+		project.getGradle().getTaskGraph().whenReady(new Closure(null) {
```

Comment:
I think there's a `whenReady` that supports an `Action` now isn't there?

Reply:
There is, but not in Gradle 2.x, which we still support.

Reply:
There is, but not in Gradle 2.x, which we still support.